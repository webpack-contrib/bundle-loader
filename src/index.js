/*
  MIT License http://www.opensource.org/licenses/mit-license.php
  Author Tobias Koppers @sokra
*/
import loaderUtils from 'loader-utils';

export function pitch(remainingRequest) { // eslint-disable-line no-unused-vars
  const options = loaderUtils.getOptions(this) || {};

  let chunkNameParam = '';

  if (options.name) {
    const opts = {
      context: options.context || this.options.context,
      regExp: options.regExp,
    };
    const chunkName = loaderUtils.interpolateName(this, options.name, opts);
    chunkNameParam = `, ${JSON.stringify(chunkName)}`;
  }

  let result;
  if (options.lazy) {
    result = [
      'module.exports = function(cb) {\n',
      ' require.ensure([], function(require) {\n',
      '   cb(require(", loaderUtils.stringifyRequest(this, `!!${remainingRequest}`), "));\n',
      `    }${chunkNameParam});\n`,
      '}'];
  } else {
    result = [
      'var cbs = [], \n',
      ' data;\n',
      'module.exports = function(cb) {\n',
      ' if(cbs) cbs.push(cb);\n',
      ' else cb(data);\n',
      '}\n',
      'require.ensure([], function(require) {\n',
      ' data = require(", loaderUtils.stringifyRequest(this, `!!${remainingRequest}`), ");\n',
      ' var callbacks = cbs;\n',
      ' cbs = null;\n',
      ' for(var i = 0, l = callbacks.length; i < l; i++) {\n',
      '   callbacks[i](data);\n',
      ' }\n',
      `}${chunkNameParam});`];
  }
  return result.join('');
}

export default () => { };

/*
Output format:

  var cbs = [],
    data;
  module.exports = function(cb) {
    if(cbs) cbs.push(cb);
    else cb(data);
  }
  require.ensure([], function(require) {
    data = require("xxx");
    var callbacks = cbs;
    cbs = null;
    for(var i = 0, l = callbacks.length; i < l; i++) {
      callbacks[i](data);
    }
  });

*/
