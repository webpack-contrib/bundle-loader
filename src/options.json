{
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "lazy": {
      "type": "boolean"
    },
    "context": {
      "type": "string"
    },
    "regExp": {}
  },
  "additionalProperties": false
}
