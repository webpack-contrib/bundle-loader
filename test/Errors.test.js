import { pitch } from '../src';

describe('Errors', () => {
  test('Validation Error', () => {
    const err = () => pitch.call({ query: { name: 1 } });

    expect(err).toThrow();
    expect(err).toThrowErrorMatchingSnapshot();
  });
});
