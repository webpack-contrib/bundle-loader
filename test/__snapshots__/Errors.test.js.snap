// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors Validation Error 1`] = `
"Validation Error

Bundle Loader Invalid Options

options.name should be string
"
`;
